### PR TITLE
Add queue controller

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
fixes: #16

### CODE REVIEW

File: pkg/apis/helpers/helpers.go [hunk 1]
```diff
@@ -63,7 +63,7 @@ func ControlledBy(obj interface{}, gvk schema.GroupVersionKind) bool {
 	return false
 }
 
-func CreateConfigMapIfNotExist(job *vkv1.Job, kubeClients *kubernetes.Clientset, data map[string]string, cmName string) error {
+func CreateConfigMapIfNotExist(job *vkv1.Job, kubeClients kubernetes.Interface, data map[string]string, cmName string) error {
```
Comment (k82cn) [inline pkg/apis/helpers/helpers.go]: prefer to have seperate PR for such kind of cleanup.

File: pkg/controllers/queue/queue_controller.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+	glog.V(5).Infoln("Begin sync queue")
+
+	var pending, running, unknown int32
+	c.pgMutex.RLock()
+	podGroups := make([]string, len(c.podGroups[key]))
+	if c.podGroups[key] == nil {
+		glog.V(2).Infoln("queue %s has not been seen or deleted")
+		return nil
+	}
+	for pgKey := range c.podGroups[key] {
```
Comment (k82cn) [inline pkg/controllers/queue/queue_controller.go]: prefer to handle this in Cache: Queue -> Job -> PodGroup.

Comment (hzxuzhonghu) [reply pkg/controllers/queue/queue_controller.go]: I donot want to overlap with job controller.  For queue controller, it just care about the podgroups, not care about Job or anything else. 

podGroups is keyed by queue name and value is the podgroups that are using this queue.

File: pkg/apis/helpers/helpers.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/apis/helpers/helpers.go]: make sense

File: pkg/controllers/queue/queue_controller.go (hunk 2 above)
Comment (k82cn) [reply pkg/controllers/queue/queue_controller.go]: `pkg/cache` is a common component in controller, and we need to clean up jobs when queue was deleted.

Comment (hzxuzhonghu) [reply pkg/controllers/queue/queue_controller.go]: IC, then we need to record jobs too.

Comment (hzxuzhonghu) [reply pkg/controllers/queue/queue_controller.go]: But have a doubt: why would user delete their queues before jobs? Is there a use case? If there is a  malicious user who does this, we should not delete the running jobs by any reason.

Comment (k82cn) [reply pkg/controllers/queue/queue_controller.go]: "malicious user" is handled by security; if admin want to delete queue, it's a heavy work to delete all jobs before delete queue.

Comment (hzxuzhonghu) [reply pkg/controllers/queue/queue_controller.go]: For this, I can think of a more simple way to do this: I remember there will be an admission controller to validate queue existence before create jobs. Then we can also set the job's owner reference.  Then when the queue is deleted, k8s garbage collector can reclaim all children resources automatically. 

So we have two choice:

1. do it in controller, need to record all dependents relationship.

1. setting ownerreference, do nothing else, leave it to GC.

File: cmd/controllers/app/server.go [hunk 3]
```diff
@@ -72,10 +76,17 @@ func Run(opt *options.ServerOption) error {
 		return err
 	}
 
-	jobController := job.NewJobController(config)
+	// TODO: add user agent for different controllers
```
Comment (TommyLike) [inline cmd/controllers/app/server.go]: What is user agent?

File: pkg/controllers/queue/queue_controller.go (hunk 2 above)
Comment (TommyLike) [reply pkg/controllers/queue/queue_controller.go]: Personally prefer not adding this into cache even we need considering deleting jobs when queue deleted.

File: cmd/controllers/app/server.go (hunk 3 above)
Comment (hzxuzhonghu) [reply cmd/controllers/app/server.go]: It is just for http server to know the client identity.

Comment (hzxuzhonghu) [reply cmd/controllers/app/server.go]: A http header

File: pkg/controllers/queue/queue_controller.go (hunk 2 above)
Comment (k82cn) [reply pkg/controllers/queue/queue_controller.go]: > setting ownerreference, do nothing else, leave it to GC.

owerReference should be reserved for `JobGroup` :)

Comment (hzxuzhonghu) [reply pkg/controllers/queue/queue_controller.go]: OwnerReferences is an array, can be more than one.  

>owerReference should be reserved for JobGroup :)

Sorry, there is no `JobGroup` now.

Comment (k82cn) [reply pkg/controllers/queue/queue_controller.go]: not sure whether GC will check controlled owner reference.

> there is no JobGroup now.

so "reserved" it.

File: pkg/controllers/queue/queue_controller.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+	return nil
+}
+
+func (c *Controller) addQueue(obj interface{}) {
+	queue := obj.(*kbv1alpha1.Queue)
+	c.queue.Add(queue.Name)
+}
+
+func (c *Controller) updateQueue(old, new interface{}) {
```
Comment (TommyLike) [inline pkg/controllers/queue/queue_controller.go]: Why adding this while leave it empty?

Comment (hzxuzhonghu) [reply pkg/controllers/queue/queue_controller.go]: will remove

**Comment by hzxuzhonghu** [comment]
I am adding e2e

**Comment by TommyLike** [comment]
@hzxuzhonghu please update the [cluster role](https://github.com/volcano-sh/volcano/blob/9b7e9d7479cca4c170267bd7298666a66ab3b31a/installer/chart/volcano/templates/controllers.yaml#L8) to get appropriate rights to handle queue resource.

**Comment by hzxuzhonghu** [comment]
Thanks

File: pkg/controllers/queue/queue_controller.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+func (c *Controller) syncQueue(key string) error {
+	glog.V(5).Infoln("Begin sync queue")
+
+	var pending, running, unknown int32
+	c.pgMutex.RLock()
+	podGroups := make([]string, len(c.podGroups[key]))
+	if c.podGroups[key] == nil {
+		glog.V(2).Infoln("queue %s has not been seen or deleted")
```
Comment (TommyLike) [inline pkg/controllers/queue/queue_controller.go]: Incorrect of info message:
```
glog.V(2).Infoln("queue %s has not been seen or deleted")
```

**Pull request merged**